### PR TITLE
(RE-6989) Add component tag check

### PR DESCRIPTION
**Pull request by McdonaldSeanp**
This adds a component tagging check for packaging. This (for now) is only a
human check and not a barrier for any CI jobs or builds using
pl:jenkins:uber_build. This commit adds the vanagon:check_tags and
vanagon:build_all rake tasks. The latter of which should replace the human
invocation of pl:jenkins:uber_build when we ship vanagon projects. The new
vanagon:build_all task is really just a wrapper for uber_build that adds
check_tags and a prompt.

Basically this is a human checklist, meant to notify the user if something
"looks tagged". It will not prevent shipment or builds, it will simply output
what "looks tagged" and then ask if this looks correct.

### CODE REVIEW

File: lib/packaging/util/file.rb [hunk 1]
```diff
@@ -32,6 +32,16 @@ def directories(dir)
       end
     end
 
+    # Returns an array of all files with #{ext} inside #{dir}
+    def files_with_ext(dir, ext)
+      if File.directory?(dir)
+        Dir.chdir(dir) do
+          Dir.glob("*").select { |entry| File.extname(entry) == ext }
```
Comment (mckern) [inline lib/packaging/util/file.rb]: Let's talk about double splats.

```ruby
Dir.glob["#{dir}/**/*.#{ext}"]
```

Comment (mckern) [inline lib/packaging/util/file.rb]: (You can probably also lose the `chdir` unless you're trying to construct relative pathnames?)

Comment (mckern) [inline lib/packaging/util/file.rb]:
```ruby
>> dir = "tmp"
=> "tmp"
>> ext = ".string"
=> ".string"
=> nil
>> puts %(Dir.glob["#{dir}/**/*#{ext}"])
Dir.glob["tmp/**/*.string"]
>> Dir.glob("#{dir}/**/*#{ext}")
=> []
>>
```

**Comment by McdonaldSeanp** [comment]
The glob returns an empty array if I use the double splat instead of select. So I took out the chdir and directory? but left that in

File: lib/packaging/util/file.rb [hunk 2]
```diff
@@ -32,6 +32,12 @@ def directories(dir)
       end
     end
 
+    # Returns an array of all files with #{ext} inside #{dir}
+    def files_with_ext(dir, ext)
+      Dir.glob("#{dir}/*").select { |entry| File.extname(entry) == ext }
```
Comment (mckern) [inline lib/packaging/util/file.rb]: Methods that return arrays should return empty arrays if there's nothing to populate the array with. This will also return an empty array if nothing is found. This implementation holds to that (as does the double-splat in the `Dir.glob` I suggested), but this one isn't recursive (The double-splat just recurses.). You'd get the same result with `Dir.glob("#{dir}/*#{ext}")` if you don't want to recurse.

Comment (mckern) [inline lib/packaging/util/file.rb]: Follow-up on what `Dir.glob().select {}` returns when there's nothing to select:

```ruby
>> dir = "tmp"
=> "tmp"
>> File.exists? dir
=> false
>> ext = ".string"
=> ".string"
>> Dir.glob("#{dir}/*").select { |entry| File.extname(entry) == ext }
=> []
>> Dir.glob("#{dir}/**/*.#{ext}")
=> []
>> Dir.glob("#{dir}/*.#{ext}")
=> []
```

File: lib/packaging/util/version.rb [hunk 3]
```diff
@@ -2,6 +2,7 @@
 
 module Pkg::Util::Version
   class << self
+    require 'json'
```
Comment (mckern) [inline lib/packaging/util/version.rb]: I would prefer we not require in the Module body if possible. `json` is part of Stdlib, so it's safe to require practically anywhere. Outside of the Module body is a good start.

File: lib/packaging/util/version.rb [hunk 4]
```diff
@@ -311,5 +312,32 @@ def versionbump(workdir = nil)
       # ...and write it back on out.
       File.open(version_file, 'w') { |f| f.write contents }
     end
+
+    def report_json_tag(path_to_json)
+      json_data = load_from_json(path_to_json)
+      output_string = " tagged at: " + json_data["ref"].to_s
+      output_string = output_string.ljust(60)
```
Comment (mckern) [inline lib/packaging/util/version.rb]: Don't use hard justifications. Start by making this output into something simple and machine readable, so you can wrap a formatter around it later if it matters.

File: lib/packaging/util/version.rb [hunk 5]
```diff
@@ -311,5 +312,32 @@ def versionbump(workdir = nil)
...
+      if /refs\/tags\/[0-9]*\.[0-9]*\.[0-9]*/.match(json_data["ref"].to_s)
+        output_string += " Tagged? [ Yes ]"
+      else
+        output_string += " Tagged? [ No  ]"
+      end
+    end
```
Comment (mckern) [inline lib/packaging/util/version.rb]: This function is busy and it's doing too much(matching/processing data, formatting & output). Let's unpack it piece by piece:

- Don't use hard justifications for output formatting; calculate them whenever possible and then feed those values into pre-formatted strings and `sprintf`. If you **really** need to do a lot of it, look into a dedicated CLI library like Highline.
- I'd probably start by making this output into something simple and machine readable, and then wrapping a formatter around it later if it matters. Even if you keep the extra-hardcoded formatting right now, it should probably still cut out into a method that operates over a placeholder instance variable.
- If you're wedded to solving "check this ref" with a regex, I think sometimes they read better if you wrap 'em in `%r{}` instead of `/ /` (at least when where's slashes in them). You won't have to escape forward slashes if you use the percent-defined version.
- You've got your "thing to match" and "match criteria" backwards. The plain string should come first, then match criteria:

```ruby
>> "refs/tags/0.9.10".match(/refs\/tags\/[0-9]*\.[0-9]*\.[0-9]*/)
=> #<MatchData "refs/tags/0.9.10">`)
>>
```

- It looks like you're really string matching (not pulling pieces out of the ref) so `#match` is heavyweight. Consider `#include?`:

```ruby
# Actual Leatherman ref from the pe-client-tools' leatherman.json; note that it's not
# fully qualified like the refs we use in puppet-agent. Git will accept a branch name,
# a tag, a fully qualified ref, or a SHA and we should strive to be accommodating.
>> ref = "0.6.2"
=> "0.6.2"
>> # Actual Leatherman refs from `git ls-remote`
>> puts %x(git ls-remote --tags https://github.com/puppetlabs/leatherman.git)
d6ebb9c49122fa6bd9d1b243ebe46dc031ab080e  refs/tags/0.6.2
565f1e811c53fc6917441fc4dcf8ea3cbb4816aa  refs/tags/0.6.2^{}
929098f3ebde88f8eec8e820b995e80f93620df2  refs/tags/0.6.3
4017302d316c7ce67a3f85c8b0b527c668ee4052  refs/tags/0.6.3^{}
>> >> "refs/tags/0.6.3".include? ref
=> false
>> "refs/tags/0.6.2".include? ref
=> true
>>
```
- Since the JSON data seems to be read-only, I'd consider wrapping the loading of JSON data into a function tied to an instance variable. Then you can memoize the instance variable and treat the function the same way you'd treat a local variable:

```ruby
def json_data
  # Try to load from an Env. variable file location first,
  # then from a default location
  @json_data ||= load_json_from_file(ENV["SOMEFILE"] || "somefile.json")
end

def report_json_tag
  output_string = " tagged at: #{json_data["ref"]}"
  ### snip
```

### But now that you hate me with all of your might!

Consider actually parsing the ref. I just wrote some code for Vanagon to do that (for the Git repos used as component sources), and it'll probably be a little more reusable than this. It's probably not drop-in perfect for your needs, but here's an example I hacked out of the pending work:

```ruby
class Git
  attr_reader :address, :ref, :ref_name, :ref_type, :branch_name

  # A SHA1 sum is 20 characters long, but Git will match on
  # the first ~8 or so. And 8 is long enough for fun test sums
  # like 'cafebeef' or 'deadfeed`.
  SHA1 = /[0-9A-F]{8,20}/i

  def initialize(address, reference)
    @address = address
    @ref = reference
    parse_ref!
  end

  def parse_ref!
    _, @ref_type, @ref_name = ref.split('/', 3)
    ref_type
  end

  def ref_type
    @ref_type ||= mangle_ref
  end

  def mangle_ref
    sha? ? "sha" : "tags"
  end

  def branch_name
    branch? ? ref_name : nil
  end

  def ref?
    return false if sha?
    !!(ref_type && ref_name)
  end

  def branch?
    ref_type.downcase == "heads"
  end

  def tag?
    ref_type.downcase == "tags"
  end

  def sha?
    !!(ref =~ SHA1)
  end
end
```

Now, brace yourself for the next novella I'm writing for the `load_from_json` function.

File: lib/packaging/util/version.rb [hunk 6]
```diff
@@ -311,5 +312,32 @@ def versionbump(workdir = nil)
...
+
+    # Loads and parses json from a file. Will treat the keys in the
+    # json as methods to invoke on the component in question
```
Comment (mckern) [inline lib/packaging/util/version.rb]: Looks like Vanagon copypasta

File: lib/packaging/util/version.rb [hunk 7]
```diff
@@ -311,5 +312,32 @@ def versionbump(workdir = nil)
...
+    #
+    # @param file [String] Path to the json file
+    # @raise [RuntimeError] exceptions are raised if there is no file, if it refers to methods that don't exist, or if it does not contain a Hash
+    def load_from_json(file)
+      if File.exists?(file)
```
Comment (mckern) [inline lib/packaging/util/version.rb]: Skip this. Just try to parse the data. It'll raise `Errno::ENOENT: No such file or directory`, which is the right error for this situation.

File: lib/packaging/util/version.rb [hunk 8]
```diff
@@ -311,5 +312,32 @@ def versionbump(workdir = nil)
...
+        data = JSON.parse(File.read(file))
+        raise "Hash required. Got '#{data.class}' when parsing '#{file}'" unless data.is_a?(Hash)
```
Comment (mckern) [inline lib/packaging/util/version.rb]: This should probably be `raise RuntimeError, blahblahblah`. It's a little more specific than plain old `StandardError`.

File: lib/packaging/util/version.rb [hunk 9]
```diff
@@ -311,5 +312,32 @@ def versionbump(workdir = nil)
...
+      else
+        fail "Cannot load component data from '#{file}'. It does not exist."
```
Comment (mckern) [inline lib/packaging/util/version.rb]: If you lose the redundant sanity check at the top of this method, you can ditch this too.

File: lib/packaging/util/version.rb [hunk 10]
```diff
@@ -311,5 +312,32 @@ def versionbump(workdir = nil)
...
+      end
+      data
```
Comment (mckern) [inline lib/packaging/util/version.rb]: As goes the redundant sanity check, so goes the explicit return value.

File: tasks/jenkins_dynamic.rake [hunk 11]
```diff
@@ -15,9 +15,25 @@ namespace :pl do
   namespace :jenkins do
     desc "Dynamic Jenkins UBER build: Build all the things with ONE job"
     task :uber_build, [:poll_interval] => "pl:fetch" do |t, args|
-      args.with_defaults(:poll_interval => 0)
+      args.with_defaults(:poll_interval => 0, :check_component_tags => false)
       poll_interval = args.poll_interval.to_i
 
+      # This is a check for vanagon component tags, and only a report.
+      # The report itself does not stop the build process.
+      if ENV['CHECK_VANAGON_COMPONENT_TAGS'] && Pkg::Config.vanagon_project
```
Comment (mckern) [inline tasks/jenkins_dynamic.rake]: Should this be moved into its own Rake task? And should that task move a lot of this formatting into dedicated output functions?

File: lib/packaging/util/version.rb (hunk 3 above)
Comment (McdonaldSeanp) [inline lib/packaging/util/version.rb]: Not sure why I did that.

File: lib/packaging/util/version.rb (hunk 5 above)
Comment (McdonaldSeanp) [inline lib/packaging/util/version.rb]: I like these ideas, PR update to come

File: lib/packaging/util/version.rb (hunk 6 above)
Comment (McdonaldSeanp) [inline lib/packaging/util/version.rb]: yup

File: lib/packaging/util/version.rb (hunk 7 above)
Comment (McdonaldSeanp) [inline lib/packaging/util/version.rb]: 👍

File: lib/packaging/util/version.rb (hunk 8 above)
Comment (McdonaldSeanp) [inline lib/packaging/util/version.rb]: 👍

File: lib/packaging/util/version.rb (hunk 9 above)
Comment (McdonaldSeanp) [inline lib/packaging/util/version.rb]: 👍

File: lib/packaging/util/version.rb (hunk 10 above)
Comment (McdonaldSeanp) [inline lib/packaging/util/version.rb]: 👍

File: tasks/jenkins_dynamic.rake (hunk 11 above)
Comment (McdonaldSeanp) [inline tasks/jenkins_dynamic.rake]: So that was my original thought, but wanted to see what others thought before adding yet another rake task to packaging. I'll add one to this PR and we can discuss

File: lib/packaging/util/version.rb (hunk 8 above)
Comment (McdonaldSeanp) [inline lib/packaging/util/version.rb]: Rubocop does not like adding the RuntimeError explicitly, it says it's redundant

Comment (mckern) [inline lib/packaging/util/version.rb]: Oh, you're right. `raise` throws `RuntimeError` by default, not `StandardError`. OK.

**Comment by McdonaldSeanp** [comment]
For some reason if I don't explicitly state the return in load_json it returns nil, so I left that in

**Comment by mckern** [comment]
@McdonaldSeanp `#load_from_json` returns `nil` without the explicit return because of the `unless`. Since the conditional in `unless` never returns anything, that's the last value to return.

It's gross, but we can probably leave it as is; building a whole new "read this junk from JSON otherwise do a thing" set of methods for Packaging isn't how you or I want to spend the afternoon. Can you add a comment explaining why you made this choice, why it works that way, and date it? I want us to know what we were thinking when we need to update this in 6-12 months.

**Comment by McdonaldSeanp** [comment]
Yeah, I'll leave that comment

File: tasks/vanagon.rake [hunk 12]
```diff
@@ -0,0 +1,24 @@
...
+
+  desc "Run component tag reports, if the report is okay, run an uber_build"
+  task :build_all, [:poll_interval] do |t, args|
+    args.with_defaults(:poll_interval => 0)
+    Rake::Task["vanagon:check_tags"].invoke
+    puts "Does this look correct?"
+    return unless Pkg::Util.ask_yes_or_no
+    Rake::Task["pl:jenkins:uber_build"].invoke(args.poll_interval)
+  end
+end
```
Comment (underscorgan) [inline tasks/vanagon.rake]: Missing newline

**Comment by underscorgan** [comment]
It looks like the commit message for this is out of date?

Also, do you have an example of what the output looks like running against puppet-agent? (just curious)

**Comment by McdonaldSeanp** [comment]
snip20160511_9

**Comment by McdonaldSeanp** [comment]
Also updated commit message

File: tasks/vanagon.rake (hunk 12 above)
Comment (McdonaldSeanp) [inline tasks/vanagon.rake]: fixed

File: lib/packaging/util/version.rb [hunk 13]
```diff
@@ -311,5 +312,35 @@ def versionbump(workdir = nil)
...
+    def report_json_tags(path_to_json)
+      json_data = Pkg::Util::Misc.load_from_json(path_to_json)
+      puts "component: " + File.basename(json_data["url"])
+      puts "tag: " + json_data["ref"].to_s
+      if tagged?(json_data["ref"].to_s)
+        tagged = "Tagged? [ Yes ]"
+      else
+        tagged = "Tagged? [ No  ]"
+      end
+      puts format("%70s", tagged)
```
Comment (mckern) [inline lib/packaging/util/version.rb]: You know that `COLUMNS` is (almost always) an environment variable in your terminal, right? You can use them to dynamically calculate terminal width and figure out how many spaces **AND** asterisks to print.

```ruby
>> spaces = "\s" * (ENV["COLUMNS"].to_i || 70)
=> "                                                                                                                                        "
>> puts "#{spaces}magic!"
                                                                                                                                        magic!
=> nil
```

Comment (McdonaldSeanp) [inline lib/packaging/util/version.rb]: I did not

**Comment by McdonaldSeanp** [comment]
I cleaned up the output a little and made it not report on json files without a ref: 
snip20160512_10

**Comment by underscorgan** [comment]
This still has a false negative but not sure how easy that would be to correct for without getting false positives? (cfpropertylist)

**Comment by McdonaldSeanp** [comment]
Yeah, what would be ideal is if we could enforce a standard, like always using refs/tags/<tag>

**Comment by mckern** [comment]
@McdonaldSeanp Aaaaaand that's why I wrote you a Git class that parsed refs.

**Comment by underscorgan** [comment]
Or maybe parse for a SHA-1 ref?

**Comment by mckern** [comment]
@underscorgan that's in the class!

```ruby
class Git
  attr_reader :address, :ref, :ref_name, :ref_type, :branch_name

  # A SHA1 sum is 20 characters long, but Git will match on
  # the first ~8 or so. And 8 is long enough for fun test sums
  # like 'cafebeef' or 'deadfeed`.
  SHA1 = /[0-9A-F]{8,20}/i

  def initialize(address, reference)
    @address = address
    @ref = reference
    parse_ref!
  end

  def parse_ref!
    _, @ref_type, @ref_name = ref.split('/', 3)
    ref_type
  end

  def ref_type
    @ref_type ||= mangle_ref
  end

  def mangle_ref
    sha? ? "sha" : "tags"
  end

  def branch_name
    branch? ? ref_name : nil
  end

  def ref?
    return false if sha?
    !!(ref_type && ref_name)
  end

  def branch?
    ref_type.downcase == "heads"
  end

  def tag?
    ref_type.downcase == "tags"
  end

  def sha?
    !!(ref =~ SHA1)
  end
end
```

**Comment by McdonaldSeanp** [comment]
Yeah, ryan had already done that, the reason I decided to leave in the regex in lieu of using the class was from a conversation I had with rob about enforcing the refs/tags/#{tag} bit. Whatever is fine with me but I liked the idea of parsing for less things rather than looking for more things.

**Comment by McdonaldSeanp** [comment]
I'll redo this with the class, gimme a minute

**Comment by mckern** [comment]
I think enforcing a ref standard is going to blow up in our faces. It's orthogonal to Git's workflow (poke around in the Git docs for the term "commitish"), and it might limit what we need to look for but it's also going to require writing tooling to enforce it. I'd rather just write tooling to parse the valid options, make reasonable assumptions about what *kind* of commitish a ref is, and then let Git do the heavy lifting.

**Comment by McdonaldSeanp** [comment]
Again, totally fine with me. It was all just based on a conversation.

**Comment by McdonaldSeanp** [comment]
New set of changes, new output: 
snip20160512_11

**Comment by McdonaldSeanp** [comment]
I added some spec tests to the git_tag class and made everything a little shinier should be pretty close at this point

File: lib/packaging/util/git_tags.rb [hunk 14]
```diff
@@ -0,0 +1,71 @@
...
+      ref_parts = ref.split('/', 3)
+      @ref_name = ref_parts.pop
+      @ref_type = ref_parts.pop
+    end
+
+    # Fetch the full ref using ls-remote, this should raise an error if it returns non-zero
+    # because that means this ref doesn't exist in the repo
+    def fetch_full_ref
+      Pkg::Util::Execution.ex("#{GIT} ls-remote --tags --heads --exit-code #{address} #{ref}").split.last
+    rescue
```
Comment (mckern) [inline lib/packaging/util/git_tags.rb]: I'd probably only rescue `RuntimeError` here; I know it's the general error class but `rescue

File: lib/packaging/util/git_tags.rb [hunk 15]
```diff
@@ -0,0 +1,71 @@
...
+    end
+
+    # Split the ref based on slashes, set ref_name and ref_type based on the last two
+    # items from the split. i.e. refs/tags/1.1.1 would return:
+    #       @ref_name => 1.1.1       @ref_type => tags
+    def split_ref(ref)
+      ref_parts = ref.split('/', 3)
+      @ref_name = ref_parts.pop
+      @ref_type = ref_parts.pop
+    end
```
Comment (mckern) [inline lib/packaging/util/git_tags.rb]: As originally written this used parallel assignment and returned an array. As written now it uses serial assignment and returns a string.

```ruby
>> def split_ref_pop(ref)
|      ref_parts = ref.split('/', 3)
|      @ref_name = ref_parts.pop
|      @ref_type = ref_parts.pop
|  end
=> :split_ref_pop
>>
>> def split_ref_multiple(ref)
|      _, @ref_type, @ref_name = ref.split('/', 3)
|  end
=> :split_ref_multiple
>> ref = "refs/tags/1.4.5"
=> "refs/tags/1.4.5"
>> split_ref_multiple(ref)
=> [
    [0] "refs",
    [1] "tags",
    [2] "1.4.5"
]
>> @ref
@ref_name  @ref_type
>> @ref_type
=> "tags"
>> @ref_name
=> "1.4.5"
>> @ref_type = nil
=> nil
>> @ref_name = nil
=> nil
>> split_ref_pop(ref)
=> "tags"
>> ref = "refs/tags"
=> "refs/tags"
>> split_ref_pop(ref)
=> "refs"
>> split_ref_multiple(ref)
=> [
    [0] "refs",
    [1] "tags"
]
>>
```

Is this desirable? Should we enforce any kind of constraint with `git check-ref-format` before we try to split it?

```bash
mckern@flexo Desktop $ git check-ref-format --normalize "ref/heads/master"
ref/heads/master
mckern@flexo Desktop $ echo $?
0
mckern@flexo Desktop $ git check-ref-format --normalize "master"
mckern@flexo Desktop $ echo $?
1
mckern@flexo Desktop $ 
```

**Comment by mckern** [comment]
I could nitpick this some more, but I'm generally 👍 for the sake of getting it in and getting us using it. I've got some questions up, but consider them non-blocking.

File: lib/packaging/util/git_tags.rb (hunk 15 above)
Comment (McdonaldSeanp) [inline lib/packaging/util/git_tags.rb]: Yeah I think it's desirable. The problem with the parallel assignment is both "heads/master" and "refs/heads/master" are acceptable ref formats. So with the first option, the parallel assignment will incorrectly return ref_name as nil and ref_type as "master". what we could do if we didn't just make the function private (which I was thinking of doing) Is we could just force the return value as: [ref_name, ref_type] or something like that.

File: lib/packaging/util/git_tags.rb (hunk 14 above)
Comment (McdonaldSeanp) [inline lib/packaging/util/git_tags.rb]: SGTM

File: lib/packaging/util/git_tags.rb (hunk 15 above)
Comment (McdonaldSeanp) [inline lib/packaging/util/git_tags.rb]: So I did the above (added the explicit return of [ref_type, ref_name] ) so it would have a nice return value, I think that solves this problem
